README.md
# iBark

[![Views](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2Fuplau%2Fibark&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=Views&edge_flat=false)](https://hits.seeyoufarm.com)
[![GitHub workflow status](https://github.com/uplau/ibark/actions/workflows/cicd.yaml/badge.svg)](https://github.com/uplau/ibark/actions/workflows/cicd.yaml)
[![GitHub release (with filter)](https://img.shields.io/github/v/release/uplau/ibark)](https://github.com/uplau/ibark/releases/latest)
![GitHub all releases](https://img.shields.io/github/downloads/uplau/ibark/total)
[![License](https://img.shields.io/badge/license-MIT%2FApache--2.0-blue.svg)](./LICENSE-MIT)

iBark is a fully featured [Bark](https://github.com/Finb/Bark) cross-platform command line tool written in Rust.

From now on, easily and securely send notifications to your apple devices from the terminal or web interface.

![demo](./.assets/demo.svg?raw=true)

## Table of contents

- [Features](#features)
- [Quick start](#quick-start)
- [Install](#install)
- [Usage](#usage)
- [Configuration](#configuration)
- [Example](#example)
- [Contributing](#contributing)
- [Contributors](#contributors)
- [License](#license)

## Features

- [x] Ease of use in mind
- [x] Shell completion `bash` `elvish` `fish` `powershell` `zsh`
- [x] Choose your `Bark` remote
- [x] Request remote `healthz` `info` `ping`
- [x] Send once notification
- [x] Specifying multiple devices to send
- [x] Support for remote basic-auth
- [x] Support end-to-end encryption
- [ ] `WIP` Web interface
- [ ] `WIP` Send template
- [ ] `WIP` Send scheduler

## Quick start

```bash
# Send once notification
$ your_device_key=""

# Simple
$ ibark send "d://${your_device_key}" -c 't=iBark 💗' -c 'b=Hello 👋\nWorld🌍'

# URL scheme
$ ibark send "d://${your_device_key}" -c 't=iBark 💗' -c 'b=Open github repo' -c 'u=https://github.com/uplau/ibark'

# Icon, available only on iOS15 or later
$ ibark send "d://${your_device_key}" -c 't=iBark 💗' -c 'b=Icon' -c 'i=https://cdn.jsdelivr.net/gh/walkxcode/dashboard-icons@master/png/apple.png'

# See more at https://github.com/Finb/bark-server/blob/master/docs/API_V2.md#push

# Request remote
$ ibark healthz
$ ibark info
$ ibark ping

# Just dump data, will not execute
$ ibark <COMMAND> -D
$ ibark <COMMAND> -DD
```

## Install

Download binaries archive from [Github releases](https://github.com/uplau/ibark/releases), unzip to your `$PATH`.

- For `Unix`, the idiomatic path is `/usr/local/bin`

## Usage

Any time run `ibark <-h|--help>` or `ibark <COMMAND> <-h|--help>`, should be obvious.

> `-h`: short help
> `--help`: verbose help

```bash
iBark is a fully featured Bark cross-platform command line tool written in Rust

Usage: ibark [OPTIONS] [COMMAND]

Commands:
  healthz  Get remote healthz
  info     Get remote info
  ping     Ping remote
  send     Send once notification
  server   [WIP]Web interface
  help     Print this message or the help of the given subcommand(s)

Options:
  -g, --gen <SHELL>         If provided, outputs the completion file for given shell
  -C, --config <PATHS>      Specify configuration files
  -D, --dump...             Just dump data, will not execute
  -R, --remote <REMOTE>     Specify remote
  -U, --user-agent <AGENT>  Specify user-agent
  -h, --help                Print help (see more with '--help')
  -V, --version             Print version
```

## Configuration

### Hierarchical structure

If you do not explicitly specify configuration files with ibark, it will in turn probe for configuration files named `preset.<ini,json,json5,toml,yml,yaml>` in the following dirs:

- global:
  - Windows: `%PUBLIC%\ibark\`
  - Unix: `/etc/ibark/`
- user: `$HOME/.config/ibark/`
- current: `./`

> **Note:** The above file does not have to exist, as it provides sensible defaults.
> If a key is specified in multiple configuration files, the values will be merged together.

> You can run `ibark <COMMAND> <-D|--dump>` to view the final configuration.

### Configuration format

> **Here are a few Special syntax:**

- `remote`: `<http|https>://[<username>:[password]@]<remote>[:port]`
  - `simple`: `https://api.day.app`
  - `port`: `https://api.day.app:65535`
  - `basic-auth`: `https://username:password@hello.world`
- `devices`: `<d|aes>://[<aes_key>:[aes_iv]@]<device_key>[<aes_key_bitlen>/<aes_mode>/<aes_padding>]`
  - `simple`: `d://zd10IOkoPAbkctTRrVPRhe`
  - `aes128cbc`: `aes://3Niw4zYjwOyYr7Bz:zOb8r84B3KrGjs1j@Uz28emqNYwUfmjolArYwyP/128/cbc/pkcs7`
  - `aes192ecb`: `aes://PO9juShj1cxZYkiNf5Bs5jxU:kOOt2GYVcqIc8EeZ@1Oto6FxeC2hTiIfNoNjofG/192/ecb/pkcs7`
  - `aes256ecb`: `aes://sN9iOmq9DniEQYLkQfieKRAZvhMOAjam:A8s7RXBFDNUbS3US@albTJWqsiDe6rjr5CsHCca/256/ecb/pkcs7`

The following is a quick overview of all settings, with `YAML` format:

> Anyone is always welcome to contribute examples in other formats
> [INI]() > [JSON]() > [TOML]()

```yaml
# fallback: https://api.day.app
remote: ...

# fallback: ibark/VERSION
user_agent: ...

devices:
  simple: ...
  aes128cbc: ...
  aes192ecb: ...
  aes256ecb: ...
  awesome_name: ...

# see more at https://github.com/Finb/bark-server/blob/master/docs/API_V2.md#push
# compatible with v1 and v2
# fallback:
#   a: 1       # or autocopy or automaticallycopy, any non-null value will be formatted as '1' to follow the upstream api
#   l: active  # or level, any null value will be formatted as 'active' to follow the upstream api
# receives first character or full key, not case sensitive, here are a few exceptions
contexts:
  bdg: 1 # or badge
  cat: "" # or category, reserved field, no use yet
  isa: 1 # or isarchive, any non-null value will be formatted as '1' to follow the upstream api

# fallback: 10
limit_conn: ...
```

## Example

> Anyone is always welcome to contribute examples

### View final configuration

```bash
$ cat preset.yaml

devices:
  awesome_name: ...

# -D   will dump related configuration
# -DD  will dump ⬆️ and configuration source
$ ibark send awesome_name -D
$ ibark send awesome_name -DD
$ ibark send d://... aes://... -D
$ ibark healthz -D
$ ibark info -D
$ ibark ping -DD
```

### Specify bark remote

```bash
$ cat preset.yaml

remote: https://hello.world

$ ibark send d://...
$ ibark send d://... -R 'https://hi.world'

# basic-auth
$ ibark send d://... -R 'https://username:password@hi.world'
```

### Send to multiple devices

```bash
$ cat preset.yaml

devices:
  simple: ...
  awesome_name: ...

$ ibark send simple awesome_name d://.... aes://...
```

### End-to-End encryption

```bash
$ cat preset.yaml

devices:
  aes128cbc: aes://3Niw4zYjwOyYr7Bz:zOb8r84B3KrGjs1j@Uz28emqNYwUfmjolArYwyP/128/cbc/pkcs7
  aes192ecb: aes://PO9juShj1cxZYkiNf5Bs5jxU:kOOt2GYVcqIc8EeZ@1Oto6FxeC2hTiIfNoNjofG/192/ecb/pkcs7
  aes256ecb: aes://sN9iOmq9DniEQYLkQfieKRAZvhMOAjam:A8s7RXBFDNUbS3US@albTJWqsiDe6rjr5CsHCca/256/ecb/pkcs7

$ ibark send aes128cbc aes192ecb aes256ecb d://...
```

### Fixed contexts

```bash
$ cat preset.yaml

contexts:
  icon: ...
  group: ...
  sound: ...

devices:
  awesome_name: ...

# will be used when contexts are not explicitly specified
$ ibark send awesome_name -D

# if you do not need these presets, you need to override them explicitly
$ ibark send awesome_name -c 'i=' -c 'g=other_group' -c 's=' -D
```

### Shell completion

```bash
# for other shells, you should be able to find relevant information from the Internet
# bash idiomatic path: /etc/bash_completion.d/
$ ibark -g bash > /etc/bash_completion.d/completion_ibark
$ source /etc/bash_completion.d/completion_ibark
```

## Contributing

This repository was created with [rust-template](https://github.com/uplau/rust-template).

See the [contributing guidelines](./CONTRIBUTING.md) for more information.

## Contributors

<a href="https://github.com/uplau/ibark/graphs/contributors">
<img src="https://contrib.rocks/image?repo=uplau/ibark&max=400&columns=20" />
</a>

## License

iBark is licensed under either of the following, at your option:

- [MIT License](./LICENSE-MIT)
- [Apache-2.0 License](./LICENSE-APACHE)


CHANGELOG.md
# Changelog

## 0.1.0 (2023-07-27)


### Features

* **core::app:** add func ([d4435aa](https://github.com/uplau/ibark/commit/d4435aa84ee9ec6b0234872f49f8453c41d498c4))
* **core::bark:** add func and tests ([c7a105a](https://github.com/uplau/ibark/commit/c7a105a2775a650807bf093a94a047f1bb33789e))
* **core::bark:** print warn bark_context_category ([7e08467](https://github.com/uplau/ibark/commit/7e08467b26f7b17730b7c56a6aa5a72445109fd0))
* **core::cli:** add func and tests ([79f4c62](https://github.com/uplau/ibark/commit/79f4c62c3eb6860ea1055400005c898711c49311))
* **core::cmd:** add func ([582a94e](https://github.com/uplau/ibark/commit/582a94e0808b7da156bb58f23b1fa38f3981d6b1))
* **core::conf:** add func and tests ([67b9b01](https://github.com/uplau/ibark/commit/67b9b01cda814d4f2125d01c2c5de3ad6434a31a))
* **core::healthz:** add func and tests ([149c6b0](https://github.com/uplau/ibark/commit/149c6b01c739110d60103df11c911f51e5512d09))
* **core::info:** add func and tests ([f382921](https://github.com/uplau/ibark/commit/f38292195510b99392d70cc23a11e75effa0eead))
* **core::misc:** merge healthz info ping ([a4a52d1](https://github.com/uplau/ibark/commit/a4a52d1b03180be73cc2b2b1f4f7180dfa947021))
* **core::ping:** add func and tests ([29b2bb7](https://github.com/uplau/ibark/commit/29b2bb75a45b3ea6258b5e80a1b06b6c15dedb8e))
* **core::send:** add func and tests ([f23e13e](https://github.com/uplau/ibark/commit/f23e13ef6d53d64bff95e61f567e2a78a560a7b8))
* **core::send:** use request once error ([1fcf415](https://github.com/uplau/ibark/commit/1fcf41593706ff880bfa4ac9272d595bbb9875d5))
* **core:app:** use clap_complete ([9260038](https://github.com/uplau/ibark/commit/92600381ca1e3d4ac705bfc648c01e4c4b599944))
* **macros:** add rules and tests ([5305d30](https://github.com/uplau/ibark/commit/5305d306144af0f0e68557922edc73c26e89c506))
* **util:** add func and tests ([9c0811b](https://github.com/uplau/ibark/commit/9c0811b608c483c4d3fbcc60f522a2f33d4d0d17))


### Bug Fixes

* **core::cli::Output:** use stderr warn ([0104151](https://github.com/uplau/ibark/commit/010415117e232295117e29da3d472249a1974fc3))
* **core::send:** postpone output exec ([0100759](https://github.com/uplau/ibark/commit/0100759322b267aa65aa1ad860944b1a1dac32c4))
* **core:** cargo fix ([24a6fce](https://github.com/uplau/ibark/commit/24a6fcebd0acfe0ea8b271c2fe3207aa643a0c4f))

## 0.1.0 (2023-07-21)


### Features

* **core::app:** add func ([d4435aa](https://github.com/uplau/ibark/commit/d4435aa84ee9ec6b0234872f49f8453c41d498c4))
* **core::bark:** add func and tests ([c7a105a](https://github.com/uplau/ibark/commit/c7a105a2775a650807bf093a94a047f1bb33789e))
* **core::bark:** print warn bark_context_category ([7e08467](https://github.com/uplau/ibark/commit/7e08467b26f7b17730b7c56a6aa5a72445109fd0))
* **core::cli:** add func and tests ([79f4c62](https://github.com/uplau/ibark/commit/79f4c62c3eb6860ea1055400005c898711c49311))
* **core::cmd:** add func ([582a94e](https://github.com/uplau/ibark/commit/582a94e0808b7da156bb58f23b1fa38f3981d6b1))
* **core::conf:** add func and tests ([67b9b01](https://github.com/uplau/ibark/commit/67b9b01cda814d4f2125d01c2c5de3ad6434a31a))
* **core::healthz:** add func and tests ([149c6b0](https://github.com/uplau/ibark/commit/149c6b01c739110d60103df11c911f51e5512d09))
* **core::info:** add func and tests ([f382921](https://github.com/uplau/ibark/commit/f38292195510b99392d70cc23a11e75effa0eead))
* **core::misc:** merge healthz info ping ([a4a52d1](https://github.com/uplau/ibark/commit/a4a52d1b03180be73cc2b2b1f4f7180dfa947021))
* **core::ping:** add func and tests ([29b2bb7](https://github.com/uplau/ibark/commit/29b2bb75a45b3ea6258b5e80a1b06b6c15dedb8e))
* **core::send:** add func and tests ([f23e13e](https://github.com/uplau/ibark/commit/f23e13ef6d53d64bff95e61f567e2a78a560a7b8))
* **core::send:** use request once error ([1fcf415](https://github.com/uplau/ibark/commit/1fcf41593706ff880bfa4ac9272d595bbb9875d5))
* **core:app:** use clap_complete ([9260038](https://github.com/uplau/ibark/commit/92600381ca1e3d4ac705bfc648c01e4c4b599944))
* **macros:** add rules and tests ([5305d30](https://github.com/uplau/ibark/commit/5305d306144af0f0e68557922edc73c26e89c506))
* **util:** add func and tests ([9c0811b](https://github.com/uplau/ibark/commit/9c0811b608c483c4d3fbcc60f522a2f33d4d0d17))


### Bug Fixes

* **core::cli::Output:** use stderr warn ([0104151](https://github.com/uplau/ibark/commit/010415117e232295117e29da3d472249a1974fc3))
* **core::send:** postpone output exec ([0100759](https://github.com/uplau/ibark/commit/0100759322b267aa65aa1ad860944b1a1dac32c4))
* **core:** cargo fix ([24a6fce](https://github.com/uplau/ibark/commit/24a6fcebd0acfe0ea8b271c2fe3207aa643a0c4f))


src/core/bark.rs
use anyhow::{anyhow, Context};
use std::collections::HashMap;

/*
* Contributors should not print `Remote` and `Device` directly in release, see dump()
* Debug is not implemented just because of development
*/

pub struct Contexts;
impl Contexts {
    pub fn verify(i: HashMap<String, String>) -> anyhow::Result<HashMap<String, String>> {
        let mut update = HashMap::with_capacity(i.len());
        for (k, v) in i.into_iter() {
            if v.is_empty() {
                continue;
            }

            match k.to_lowercase().as_str() {
                "a" | "autocopy" | "automaticallycopy" => {
                    let v = "1"; // must be 1
                    update.insert("automaticallyCopy".into(), v.into()); // v2
                    update.insert("autoCopy".into(), v.into()); // v1
                }
                "bdg" | "badge" => match v.parse::<i32>() {
                    Ok(_) => {
                        update.insert("badge".into(), v);
                    }
                    Err(_) => return Err(anyhow!("bark_context_badge `{v}` not a number")),
                },
                "b" | "body" => {
                    // raw \n = \\n
                    update.insert("body".into(), v.replace("\\n", "\n"));
                }
                "c" | "copy" => {
                    // raw \n = \\n
                    update.insert("copy".into(), v.replace("\\n", "\n"));
                }
                "cat" | "category" => {
                    super::cli::Output::warn("bark_context_category is not used yet");
                    // update.insert("category".into(), v);
                }
                "g" | "group" => {
                    update.insert("group".into(), v);
                }
                "i" | "icon" => {
                    update.insert("icon".into(), v);
                }
                "isa" | "isarchive" => {
                    update.insert("isArchive".into(), "1".into());
                }
                "l" | "level" => {
                    let m = ["active", "timeSensitive", "passive"];
                    if m.contains(&v.as_str()) {
                        update.insert("level".into(), v);
                    } else {
                        return Err(anyhow!(
                            "bark_context_level `{v}` not match `{}`",
                            m.join("|")
                        ));
                    }
                }
                "s" | "sound" => {
                    // https://github.com/Finb/Bark/tree/master/Sounds
                    update.insert("sound".into(), v);
                }
                "t" | "title" => {
                    update.insert("title".into(), v.replace("\\n", "\n"));
                }
                "u" | "url" => {
                    update.insert("url".into(), v);
                }
                _ => return Err(anyhow!("unsupported bark_context `{k}`")),
            }
        }
        Ok(update)
    }
}

// maybe not Debug
#[derive(Debug, Default)]
pub struct Device {
    scheme: String,
    key: String,
    aes_mode: Option<String>,
    aes_key: Option<Vec<u8>>,
    aes_iv: Option<Vec<u8>>,
    aes_padding: Option<String>, // temporarily ignored
}

impl Device {
    pub fn find_merge(s: &HashMap<String, String>, f: Vec<String>) -> HashMap<String, String> {
        let mut merge = HashMap::with_capacity(f.len());
        for find in f.into_iter() {
            if find.is_empty() {
                continue;
            }
            if s.contains_key(&find) {
                if merge.contains_key(&find) {
                    continue;
                }
                let value = s.get(&find).unwrap().to_string();
                merge.insert(find, value);
            } else {
                let key = crate::util::hash_hex_string(&find);
                if merge.contains_key(&key) {
                    continue;
                }
                merge.insert(key, find);
            }
        }
        merge
    }

    pub fn new(input: &str) -> anyhow::Result<Self> {
        let u = url::Url::parse(input).with_context(|| Self::parse_context("input", input))?;

        Self::default()
            .parse_scheme(u.scheme())?
            .parse_key(u.host_str().unwrap_or_default())?
            .parse_aes(&u)
    }

    pub fn dump(input: &str) -> anyhow::Result<String> {
        let _self = Self::new(input)?;

        match _self.scheme.as_str() {
            "d" | "aes" => {
                let key_len = _self.key.len();
                let key_secret = "*".repeat(key_len);

                if _self.scheme == "d" {
                    Ok(format!("{}://{key_secret}", _self.scheme))
                } else {
                    let aes_key_len = _self.aes_key.unwrap().len();
                    let aes_secret = format!("{}:{}@", "*".repeat(aes_key_len), "*".repeat(16));

                    let path_segments = format!(
                        "{}/{}/{}",
                        aes_key_len * 8,
                        _self.aes_mode.unwrap(),
                        _self.aes_padding.unwrap()
                    );

                    Ok(format!(
                        "{}://{aes_secret}{key_secret}/{path_segments}",
                        _self.scheme
                    ))
                }
            }
            _ => unreachable!(),
        }
    }

    pub fn new_request(
        input: &str,
        client: &reqwest::Client,
        remote: &str,
        contexts: &HashMap<String, String>,
    ) -> anyhow::Result<reqwest::RequestBuilder> {
        let _self = Self::new(input)?;
        let mut request = client.post(format!("{remote}/{}", _self.key));

        match _self.scheme.as_str() {
            "d" => {
                request = request.form(contexts);
            }
            "aes" => {
                let data = serde_json::to_vec(contexts)?;

                let cipher = match _self.aes_mode.unwrap().as_str() {
                    "cbc" => {
                        let bitlen = _self.aes_key.as_deref().unwrap().len() * 8;
                        match bitlen {
                            128 => openssl::symm::Cipher::aes_128_cbc(),
                            192 => openssl::symm::Cipher::aes_192_cbc(),
                            256 => openssl::symm::Cipher::aes_256_cbc(),
                            _ => {
                                unreachable!()
                            }
                        }
                    }
                    "ecb" => {
                        let bitlen = _self.aes_key.as_deref().unwrap().len() * 8;
                        match bitlen {
                            128 => openssl::symm::Cipher::aes_128_ecb(),
                            192 => openssl::symm::Cipher::aes_192_ecb(),
                            256 => openssl::symm::Cipher::aes_256_ecb(),
                            _ => {
                                unreachable!()
                            }
                        }
                    }
                    _ => unreachable!(),
                };

                let ciphertext_base64 = openssl::base64::encode_block(&openssl::symm::encrypt(
                    cipher,
                    &_self.aes_key.unwrap(),
                    _self.aes_iv.as_deref(),
                    &data,
                )?);

                let params = HashMap::from([("ciphertext", ciphertext_base64.as_str())]);
                request = request.form(&params);
            }
            _ => unreachable!(),
        }

        Ok(request)
    }

    fn parse_context(t: &str, input: &str) -> String {
        format!("parse bark_device_{t} `{input}` failed")
    }

    fn parse_scheme(mut self, scheme: &str) -> anyhow::Result<Self> {
        match scheme {
            "d" | "aes" => self.scheme = scheme.to_string(),
            _ => return Err(anyhow!("unsupported bark_device_scheme `{scheme}`")),
        }
        Ok(self)
    }

    fn parse_key(mut self, key: &str) -> anyhow::Result<Self> {
        let len = key.len();
        let len_check: usize = 22;
        if len < len_check {
            let secret = "*".repeat(len);
            return Err(anyhow!("len `{len}` < {len_check}"))
                .with_context(|| Self::parse_context("key", &secret));
        }
        self.key = key.to_string();
        Ok(self)
    }

    fn parse_aes(mut self, u: &url::Url) -> anyhow::Result<Self> {
        match self.scheme.as_str() {
            "aes" => {
                let path_segments: Vec<_> = u
                    .path_segments()
                    .with_context(|| "path segments is none")
                    .with_context(|| {
                        Self::parse_context("aes", "aes://aes_key:aes_iv@device_key/^here")
                    })?
                    .collect();

                let count = path_segments.len();
                if count < 3 {
                    return Err(anyhow!("count `{count}` < 3"))
                        .with_context(|| Self::parse_context("aes_path_segments", u.path()));
                }

                // let aes_key_bitlen = *path_segments.get(0).unwrap();
                let aes_key_bitlen = *path_segments.first().unwrap();
                match aes_key_bitlen {
                    "128" | "192" | "256" => {
                        let aes_key_bitlen = aes_key_bitlen.parse::<usize>()?;
                        let input_aes_key_bitlen = u.username().len() * 8;

                        if input_aes_key_bitlen != aes_key_bitlen {
                            return Err(anyhow!(
                                "bitlen `{input_aes_key_bitlen}` not match `{aes_key_bitlen}`"
                            ))
                            .with_context(|| {
                                Self::parse_context(
                                    "aes_key",
                                    &format!("aes://^here:aes_iv@device_key/{aes_key_bitlen}"),
                                )
                            });
                        }

                        let aes_iv = u.password().unwrap_or_default();
                        let input_aes_iv_bitlen = aes_iv.len() * 8;
                        if input_aes_iv_bitlen != 128 {
                            return Err(anyhow!("bitlen `{input_aes_iv_bitlen}` not match `128`"))
                                .with_context(|| {
                                    Self::parse_context("aes_iv", "aes://aes_key:^here")
                                });
                        }

                        self.aes_key = Some(hex::decode(hex::encode(u.username()))?);
                        self.aes_iv = Some(hex::decode(hex::encode(aes_iv))?);
                    }
                    _ => {
                        return Err(anyhow!(
                            "unsupported bark_device_aes_key_bitlen `{aes_key_bitlen}`"
                        ));
                    }
                }

                let aes_mode = *path_segments.get(1).unwrap();
                match aes_mode {
                    "cbc" | "ecb" => self.aes_mode = Some(aes_mode.to_string()),
                    _ => {
                        return Err(anyhow!("unsupported bark_device_aes_mode `{aes_mode}`"));
                    }
                }

                let aes_padding = *path_segments.get(2).unwrap();
                match aes_padding {
                    "pkcs7" => self.aes_padding = Some(aes_padding.to_string()),
                    _ => {
                        return Err(anyhow!(
                            "unsupported bark_device_aes_padding `{aes_padding}`"
                        ));
                    }
                }

                Ok(self)
            }
            _ => Ok(self),
        }
    }
}

pub struct Remote;
impl Remote {
    pub fn verify(remote: &str) -> anyhow::Result<()> {
        let url = url::Url::parse(remote)
            .with_context(|| format!("parse bark_remote `{remote}` failed"))?;

        let scheme = url.scheme();
        if !["http", "https"].contains(&scheme) {
            return Err(anyhow!("unsupported bark_remote_scheme `{scheme}`"));
        }
        Ok(())
    }

    pub fn dump(remote: &str) -> anyhow::Result<String> {
        let mut url = url::Url::parse(remote)?;
        if url.password().is_some() {
            let len = url.password().unwrap().len();
            url.set_password(Some(&"*".repeat(len))).unwrap();
        }
        Ok(url.into())
    }

    pub fn scheme_host_port(remote: &str) -> anyhow::Result<String> {
        let url = url::Url::parse(remote)?;
        let host = url.host_str().unwrap_or("unknown_host");
        let port = url.port_or_known_default().unwrap_or(0);
        Ok(format!("{}://{host}:{port}", url.scheme()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::util::tests::*;

    #[test]
    fn test_contexts() -> anyhow::Result<()> {
        dbg!(Contexts::verify(
            crate::hash_map! {
                "a"=>"true",
                "t"=>"iBark 💗",
                "b"=>r#"Hello 👋\nWorld🌍"#,
                "bdg"=>"100",
                "l"=>"active",
                "cat"=>"test",
                "g"=>""
            }
            .into_iter()
            .map(|(k, v)| (k.into(), v.into()))
            .collect(),
        )?);
        Ok(())
    }

    #[test]
    fn test_device() -> anyhow::Result<()> {
        let gen_args = |is_err: bool| {
            let mut args = Vec::with_capacity(5);

            args.push(format!("d://{}", random_string(22)));

            args.push(format!(
                "aes://{}:{}@{}/128/cbc/pkcs7",
                random_string(16),
                random_string(16),
                random_string(22)
            ));
            args.push(format!(
                "aes://{}:{}@{}/192/ecb/pkcs7",
                random_string(24),
                random_string(16),
                random_string(22)
            ));
            args.push(format!(
                "aes://{}:{}@{}/256/ecb/pkcs7",
                random_string(32),
                random_string(16),
                random_string(22)
            ));

            if is_err {
                args.push(random_string(random_number(0, 20) as usize));
            }
            args
        };

        let args = gen_args(false);
        dbg!(&args);
        crate::println_dash!(50);

        for input in args.clone().into_iter() {
            dbg!(Device::dump(&input)?);
            crate::println_dash!(20);
        }

        for input in args.into_iter() {
            dbg!(Device::new_request(
                input.as_ref(),
                &reqwest::Client::new(),
                &format!("https://{}.com", random_string(10)),
                &HashMap::new()
            )?);
            crate::println_dash!(20);
        }

        Ok(())
    }

    #[test]
    fn test_remote() -> anyhow::Result<()> {
        dbg!(Remote::verify("https://hello.world:65535")?);
        dbg!(Remote::dump("https://username:password@hello.world:65534")?);
        Ok(())
    }
}
